### PR TITLE
Seperate compilation

### PR DESCRIPTION
**Pull request by rashwini21**
Professor @stiber , creating this pull request for review. The performance for GPU and CPU input config files(tiny to large) deviate not more than 2 sec. Here is the google spreadsheet for the performance. 

" https://docs.google.com/spreadsheets/d/1WW1iFUQ-m6wms38ipnbmw4X_CgMmMdAKxw0CbkBFYcs/edit?usp=sharing "

### CODE REVIEW

**Review by stiber — CHANGES_REQUESTED** [review]
Mostly minor things. See individual comments.

File: CMakeLists.txt [hunk 1]
```diff
@@ -117,26 +118,78 @@ include_directories(
         ThirdParty/TinyXPath
         ThirdParty/log4cplus-2.0.7/include)
 
+
+       
+if(ENABLE_CUDA)
+        # Simulator/Core/GPUModel.cpp
+        set(cuda_sources Simulator/Edges/Neuro/AllSynapsesDeviceFuncs_d.cpp Simulator/Vertices/Neuro/AllVerticesDeviceFuncs_d.cpp Simulator/Edges/Neuro/AllSpikingSynapses_d.cpp Simulator/Edges/Neuro/AllSTDPSynapses_d.cpp Simulator/Edges/Neuro/AllDSSynapses_d.cpp Simulator/Edges/Neuro/AllDynamicSTDPSynapses_d.cpp Simulator/Vertices/Neuro/AllLIFNeurons_d.cpp Simulator/Vertices/Neuro/AllIZHNeurons_d.cpp Simulator/Vertices/Neuro/AllIFNeurons_d.cpp Simulator/Vertices/Neuro/AllSpikingNeurons_d.cpp Simulator/Connections/Neuro/ConnGrowth_d.cpp )
```
[the remark this answers is not in the document]
Comment (rashwini21) [reply CMakeLists.txt:125]: Tried 2 approaches 
1. Use Set() three times to break this long line, but cuda_sources get rewrite everytime we use set. (Not worked)
2. Break cuda_sources to cudaEdges_Source , cudaVertices_Source and cudaConnections_Source and respectively add to Vertices_Source , Edges_Source and Connections_Source but need to add all three variables together when using. (Worked but increases line of code)


Used approach 
Enter each value for cuda_Sources on different line.

File: CMakeLists.txt [hunk 2]
```diff
@@ -117,26 +118,78 @@ include_directories(
...
+        
+        #set(cuda_sources Simulator/Edges/Neuro/AllSynapsesDeviceFuncs_d.cpp Simulator/Vertices/Neuro/AllVerticesDeviceFuncs_d.cpp)
```
[the remark this answers is not in the document]
Comment (rashwini21) [reply CMakeLists.txt:127]: Removed previous comments now

File: CMakeLists.txt [hunk 3]
```diff
@@ -117,26 +118,78 @@ include_directories(
...
+        list(REMOVE_ITEM Vertices_Source "${CMAKE_CURRENT_SOURCE_DIR}/Simulator/Vertices/Neuro/AllIFNeurons_d.cpp")
+        list(REMOVE_ITEM Vertices_Source "${CMAKE_CURRENT_SOURCE_DIR}/Simulator/Vertices/Neuro/AllSpikingNeurons_d.cpp")
+        
+       
+        add_library(Vertices STATIC ${Vertices_Source})
+        
+endif()
 
 # Create Edges library
 file(GLOB_RECURSE Edges_Source Simulator/Edges/*.cpp Simulator/Edges/*.h Simulator/Edges/*.cu)
```
[the remark this answers is not in the document]
Comment (rashwini21) [reply CMakeLists.txt]: Removed .cu from the Edges_Source ,No .cu files left in Edges_Source. .cu files 2 files left in Simulator/Utils and .cu files in mersenne twister.

File: CMakeLists.txt [hunk 4]
```diff
@@ -146,7 +199,20 @@ add_library(FunctionNodes ${FunctionNodes_Source})
 file(GLOB Core_Source Simulator/Core/*.cpp Simulator/Core/*.h Simulator/Core/*.cu)
```
[the remark this answers is not in the document]
Comment (rashwini21) [reply CMakeLists.txt]: Removed .cu from the Core_Source ,No .cu files left in Core_Source. .cu files 2 files left in Simulator/Utils and .cu files in mersenne twister.

File: CMakeLists.txt [hunk 5]
```diff
@@ -178,6 +244,8 @@ include_directories(Testing/lib/GoogleTest/googletest-master/googletest/include)
 
 # Set CUDA_SEPERABLE_COMPILATION to ON for all libraries that contain .cu CUDA files
```
[the remark this answers is not in the document]
Comment (rashwini21) [reply CMakeLists.txt]: Changed Reference here

File: Simulator/Edges/Neuro/AllSynapsesDeviceFuncs.h [hunk 6]
```diff
@@ -18,7 +18,14 @@ using namespace std;
 
 #if defined(__CUDACC__)
 
-extern __device__ enumClassSynapses classSynapses_d;
+extern CUDA_CALLABLE enumClassSynapses classSynapses_d;
+extern CUDA_CALLABLE int edgSign( edgeType t );
+// extern CUDA_CALLABLE void changeSpikingSynapsesPSRDevice(AllSpikingSynapsesDeviceProperties* allEdgesDevice, const BGSIZE iEdg, const uint64_t simulationStep, const BGFLOAT deltaT);
```
[the remark this answers is not in the document]
Comment (rashwini21) [reply Simulator/Edges/Neuro/AllSynapsesDeviceFuncs.h:23]: Removed the commented out code.

File: Simulator/Vertices/Neuro/AllVerticesDeviceFuncs.h [hunk 7]
```diff
@@ -26,7 +30,7 @@
 ///  @param[in] allEdgesDevice     Pointer to Synapse structures in device memory.
 ///  @param[in] edgeIndexMap       Inverse map, which is a table indexed by an input neuron and maps to the synapses that provide input to that neuron.
 ///  @param[in] fAllowBackPropagation True if back propagaion is allowed.
-extern __global__ void advanceLIFNeuronsDevice( int totalVertices, int maxEdges, int maxSpikes, const BGFLOAT deltaT, uint64_t simulationStep, float* randNoise, AllIFNeuronsDeviceProperties* allVerticesDevice, AllSpikingSynapsesDeviceProperties* allEdgesDevice, EdgeIndexMap* edgeIndexMapDevice, bool fAllowBackPropagation );
+// extern __global__ void advanceLIFNeuronsDevice( int totalVertices, int maxEdges, int maxSpikes, const BGFLOAT deltaT, uint64_t simulationStep, float* randNoise, AllIFNeuronsDeviceProperties* allVerticesDevice, AllSpikingSynapsesDeviceProperties* allEdgesDevice, EdgeIndexMap* edgeIndexMapDevice, bool fAllowBackPropagation );
```
[the remark this answers is not in the document]
Comment (rashwini21) [reply Simulator/Vertices/Neuro/AllVerticesDeviceFuncs.h:33]: Removed here

File: Simulator/Vertices/Neuro/AllVerticesDeviceFuncs.h [hunk 8]
```diff
@@ -40,6 +44,6 @@ extern __global__ void advanceLIFNeuronsDevice( int totalVertices, int maxEdges,
 ///  @param[in] allEdgesDevice     Pointer to Synapse structures in device memory.
 ///  @param[in] edgeIndexMap       Inverse map, which is a table indexed by an input neuron and maps to the synapses that provide input to that neuron.
 ///  @param[in] fAllowBackPropagation True if back propagaion is allowed.
-extern __global__ void advanceIZHNeuronsDevice( int totalVertices, int maxEdges, int maxSpikes, const BGFLOAT deltaT, uint64_t simulationStep, float* randNoise, AllIZHNeuronsDeviceProperties* allVerticesDevice, AllSpikingSynapsesDeviceProperties* allEdgesDevice, EdgeIndexMap* edgeIndexMapDevice, bool fAllowBackPropagation );
+/// extern __global__ void advanceIZHNeuronsDevice( int totalVertices, int maxEdges, int maxSpikes, const BGFLOAT deltaT, uint64_t simulationStep, float* randNoise, AllIZHNeuronsDeviceProperties* allVerticesDevice, AllSpikingSynapsesDeviceProperties* allEdgesDevice, EdgeIndexMap* edgeIndexMapDevice, bool fAllowBackPropagation );
```
[the remark this answers is not in the document]
Comment (rashwini21) [reply Simulator/Vertices/Neuro/AllVerticesDeviceFuncs.h:47]: Removed here

File: CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (stiber) [reply CMakeLists.txt]: So, you can’t just split it over multiple lines or use a “\”? Just curious.

[the remark this answers is not in the document]
Comment (stiber) [reply CMakeLists.txt]: Never mind; that’s what you did!

**Comment by rashwini21** [comment]
Hi Professor @stiber , we can merge this branch now. Tested all clang-format changes on CPU and GPU Outputs ,no change in any output or time of running input config files. I  will now check the time change during compilation.

**Review by stiber — APPROVED** [review]

**Pull request merged**